File: client/src/models/region/ResourceNode.ts
import { TerritoryEvent } from "./TerritoryEvent";
import { TerritoryEventItem } from "./TerritoryEventItem";

export class ResourceNode extends TerritoryEvent {
  static FromParsedBody(json: any): ResourceNode {
    return new ResourceNode(
      (json["territoryEventItems"] as TerritoryEventItem[])
        .map(rni => {
          return new TerritoryEventItem(
            rni.itemName,
            rni.value,
            rni.minimumQuantity,
            rni.maximumQuantity,
          );
        }),
      json["name"],
      json["type"],
      json["description"],
      json["iconPath"],
      json["capacity"],
      json["durationSeconds"] * 1000, //  Convert seconds to ms
      json["instrumentType"],
      json["instrumentRequiredLevel"],
      json["instrumentExpectedLevel"],
    );
  }
}

File: EntityModels.Postgresql/ResourceNode.cs
using System.ComponentModel.DataAnnotations;

namespace EntityModels.Postgresql;

public partial class ResourceNode: TerritoryEvent
{
    public int Capacity { get; set; }
    public int DurationSeconds { get; set; }
    [StringLength(50)] public string InstrumentType { get; set; } = null!;
    public int InstrumentRequiredLevel { get; set; }
    public int InstrumentExpectedLevel { get; set; } //  (x2 faster for level above and 4x slower for level below)

    public ICollection<TerritoryEventItem> TerritoryEventItems { get; set; } = null!;
    public ICollection<TerritoryEventProportion> TerritoryEventProportions { get; set; } = null!;
}